packages/core-contracts/src/contracts/arks/Psm3ERC4626Ark.sol

```diff
@@ -0,0 +1,243 @@
+ *      3. Optionally stake in sUSDS vault
+ *      4. Deposit in ERC4626 vault
+ *
+ * The contract supports two modes:
+ * - Direct USDS staking: USDC -> USDS -> ERC4626 vault
+ * - sUSDS staking: USDC -> sUSDS -> ERC4626 vault
+ */
+contract Psm3ERC4626Ark is Ark {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for IERC4626;
+
+    /*//////////////////////////////////////////////////////////////
+                            STATE VARIABLES
+    //////////////////////////////////////////////////////////////*/
+    /// @notice The PSM3 contract for USDC <-> sUSDS swaps
+    IPsm3 public immutable psm;
+    /// @notice The USDS token contract
+    IERC20 public immutable usds;
+    /// @notice The sUSDS token contract
+    IERC20 public immutable susds;
+    /// @notice The ERC4626 vault contract
+    IERC4626 public immutable erc4626Vault;
+    /// @notice Whether to stake USDS in sUSDS vault
+    bool public immutable shouldStake;
+
+    /*//////////////////////////////////////////////////////////////
+                                ERRORS
+    //////////////////////////////////////////////////////////////*/
+    error InvalidPSMAddress();
+    error InvalidUSDSAddress();
+    error InvalidSUSDSAddress();
+    error InvalidGemAddress();
+
+    /*//////////////////////////////////////////////////////////////
+                            CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+    /**
+     * @notice Creates a new Psm3ERC4626Ark instance
+     * @param _psm Address of the PSM3 contract
+     * @param _usds Address of the USDS token
+     * @param _susds Address of the sUSDS token
+     * @param _erc4626Vault Address of the ERC4626 vault
+     * @param _params Ark parameters
+     */
+    constructor(
+        address _psm,
+        address _usds,
+        address _susds,
+        address _erc4626Vault,
+        ArkParams memory _params
+    ) Ark(_params) {
+        if (_psm == address(0)) revert InvalidPSMAddress();
+        if (_usds == address(0)) revert InvalidUSDSAddress();
+        if (_susds == address(0)) revert InvalidSUSDSAddress();
+        if (_erc4626Vault == address(0)) revert InvalidVaultAddress();
+
+        psm = IPsm3(_psm);
+        usds = IERC20(_usds);
+        susds = IERC20(_susds);
+        erc4626Vault = IERC4626(_erc4626Vault);
+        shouldStake = erc4626Vault.asset() == address(susds);
+
+        if (psm.usdc() != _params.asset) revert InvalidGemAddress();
+        _validateVaultAsset();
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            PUBLIC FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+    /**
+     * @notice Returns the total assets in the vault
+     * @return assets Total assets in USDC terms
+     */
+    function totalAssets() public view override returns (uint256 assets) {
+        uint256 balance = erc4626Vault.balanceOf(address(this));
+        if (balance > 0) {
```

Comment:
Is this the usual case? If in general we expect the vault to have a non-zero balance, this is actually hurting the gas optimization (as 99.99% of the time the balance we are not zero but we are still checking it)